lock/lock.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lock
+
+import (
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/infoschema/perfschema"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table"
+)
+
+// Checker uses to check tables lock.
+type Checker struct {
+	ctx sessionctx.Context
+	is  infoschema.InfoSchema
+}
+
+// NewChecker return new lock Checker.
+func NewChecker(ctx sessionctx.Context, is infoschema.InfoSchema) *Checker {
+	return &Checker{ctx: ctx, is: is}
+}
+
+// CheckTableLock uses to check table lock.
+func (c *Checker) CheckTableLock(db, table string, privilege mysql.PrivilegeType) error {
+	if db == "" && table == "" {
+		return nil
+	}
+	// Below database are not support table lock.
+	if db == infoschema.LowerName || db == perfschema.LowerName || db == mysql.SystemDB {
+		return nil
+	}
+	// check operation on database.
+	if table == "" {
+		return c.CheckLockInDB(db, privilege)
+	}
+	switch privilege {
+	case mysql.ShowDBPriv, mysql.AllPrivMask:
+		// AllPrivMask only used in show create table statement now.
+		return nil
+	case mysql.CreatePriv, mysql.CreateViewPriv:
+		if c.ctx.HasLockedTables() {
```

Comment:
It seems that the error msg is different with MySQL.
```
tidb> create table a(a int);
Query OK, 0 rows affected (0.09 sec)
tidb> lock tables a write;
Query OK, 0 rows affected (0.16 sec)
```
mysql:
```
mysql>  create table a(xx int);
ERROR 1050 (42S01): Table 'a' already exists
```
tidb:
```
tidb> create table a(xx int);
ERROR 1100 (HY000): Table 'a' was not locked with LOCK TABLES
```

Reply:
em... Because the check order is different...

Reply:
Maybe we can add a `TODO` here?

Reply:
Already add a `TODO` here. Thanks.